File: leetcode/Python/split-linked-list-in-parts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        sol, lens, n ,g  = list(), 0, 0, 0
        cur, t = head, None
        while cur:
            lens+=1
            cur=cur.next
        cur = head
        while cur:
            if n == 0:
                n = (lens//k)+1 if lens%k>0 else (lens//k)
                sol.append(cur)
            n-=1
            if n==0:
                cur.next, cur = None, cur.next
                if lens%k>0:
                    lens-=1
                g+=1
            else:
                cur=cur.next
        while g<k:
            g+=1
            sol.append(None)
        return sol
            

File: leetcode/C++/split-linked-list-in-parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        vector<ListNode*> sol;
        int lens=0,n=0,g=0;
        for(ListNode* cur=head; cur!=nullptr; cur=cur->next) ++lens;
        for(ListNode* cur=head, *t=nullptr; cur!=nullptr;){
            if(n==0) {
                n=(lens/k)+(lens%k>0);
                sol.push_back(cur);
            }
            if(--n==0) {
                t=cur->next;
                cur->next=nullptr;
                cur=t;
                if(lens%k) --lens;
                ++g;
            }
            else cur=cur->next;
        }
        for(; g<k; ++g) sol.push_back(nullptr);
        return sol;
    }
};